internal/setup/setup.go

```diff
@@ -772,3 +817,66 @@ func Select(release *Release, slices []SliceKey) (*Selection, error) {
 
 	return selection, nil
 }
+
+// pathInfoToYAML converts a PathInfo object to a yamlPath object.
+func pathInfoToYAML(pi *PathInfo) (*yamlPath, error) {
+	path := &yamlPath{
+		Mode:    yamlMode(pi.Mode),
+		Mutable: pi.Mutable,
+		Until:   pi.Until,
+		Arch:    yamlArch{List: pi.Arch},
+	}
+	switch pi.Kind {
+	case DirPath:
+		path.Dir = true
+	case CopyPath:
+		path.Copy = pi.Info
+	case TextPath:
+		path.Text = &pi.Info
+	case SymlinkPath:
+		path.Symlink = pi.Info
+	case GlobPath:
+	default:
+		return nil, fmt.Errorf("internal error: unrecognised PathInfo type: %s", pi.Kind)
+	}
+	return path, nil
+}
+
+// sliceToYAML converts a Slice object to a yamlSlice object.
+func sliceToYAML(s *Slice) (*yamlSlice, error) {
+	slice := &yamlSlice{
+		Essential: make([]string, 0, len(s.Essential)),
+		Contents:  make(map[string]*yamlPath, len(s.Contents)),
+		Mutate:    s.Scripts.Mutate,
+	}
+	for _, key := range s.Essential {
+		slice.Essential = append(slice.Essential, key.String())
+	}
+	sort.Strings(slice.Essential)
+	for path, info := range s.Contents {
+		pi := info
+		yamlPath, err := pathInfoToYAML(&pi)
+		if err != nil {
+			return nil, err
+		}
+		slice.Contents[path] = yamlPath
+	}
+	return slice, nil
+}
+
+// packageToYAML converts a Package object to a yamlPackage object.
+func packageToYAML(p *Package) (*yamlPackage, error) {
+	pkg := &yamlPackage{
+		Name:    p.Name,
+		Archive: p.Archive,
+		Slices:  make(map[string]yamlSlice, len(p.Slices)),
+	}
+	for name, slice := range p.Slices {
+		yamlSlice, err := sliceToYAML(slice)
+		if err != nil {
+			return nil, err
+		}
+		pkg.Slices[name] = *yamlSlice
+	}
+	return pkg, nil
+}
```

Comment:
800+ lines... most of this file is probably about marshaling yaml by now, which isn't great.

**After the PR is in,** let's please move the yaml-specific logic to its own file.

Reply:
Sounds good, will do!